Use case: Cohere and OpenSearch (#3123)
astronomer/docs: learn/use-case-llm-customer-feedback.md

### Project overview

This project contains two DAGs, one for the MLOps pipeline and one DAG to delete the index in OpenSearch for testing purposes. 

The [`analyze_customer_feedback` DAG](https://github.com/astronomer/use-case-llm-customer-feedback/blob/main/dags/analyze_customer_feedback.py) ingests data from the mock API and loads it into OpenSearch. The DAG then uses the Cohere API to get sentiment and embeddings for a subset of the customer feedback returned by a keyword OpenSearch query. The embeddings and sentiment analysis scores are ingested back into OpenSearch and a final query is performed to get the positive feedback most similar to a target testimonial. The DAG ends by printing the retrieved testimonial to the logs.

![Screenshot of the Airflow UI Grid view with the Graph view selected showing a successful run of the full use case DAG with 17 tasks.](/img/examples/use-case-airflow-cohere-opensearch_full_dag.png)

> The [`delete_opensearch_index` DAG](https://github.com/astronomer/use-case-llm-customer-feedback/blob/main/dags/delete_opensearch_index.py) deletes the `INDEX_TO_DELETE` in OpenSearch. This DAG is used during development to allow the the `analyze_customer_feedback` DAG to create the index from scratch. Run this DAG if you would like to start from a clean slate, for example when making changes to the index schema to adapt the project to your own use case.

Reviewer: ```suggestion
The [`delete_opensearch_index` DAG](https://github.com/astronomer/use-case-llm-customer-feedback/blob/main/dags/delete_opensearch_index.py) deletes the `INDEX_TO_DELETE` in OpenSearch. This DAG is used during development to allow the `analyze_customer_feedback` DAG to create the index from scratch. Run this DAG if you would like to start from a clean slate, for example when making changes to the index schema to adapt the project to your own use case.
```

    ![Graph view of the query section of the analyze_customer_feedback DAG.](/img/examples/use-case-airflow-cohere-opensearch_query_section.png)

- Perform sentiment analysis on relevant customer feedback and get embeddings using the Cohere API

    ![Graph view of the section in the analyze_customer_feedback DAG that performs sentiment analysis, vector embeddings with the Cohere API and loads the results back into OpenSearch.](/img/examples/use-case-airflow-cohere-opensearch_sentiment_embedding_section.png)

- Query OpenSearch for the most similar testimonial using a k-nearest neighbors (k-NN) algorithm on the embeddings and filter for positive sentiment

    ![Graph view of the final section of the analyze_customer_feedback DAG that queries OpenSearch for the most similar testimonial using k-NN on the embeddings and filter for positive sentiment.](/img/examples/use-case-airflow-cohere-opensearch_final_section.png)

Several parameters are set at the beginning of the DAG. You can adjust the number of pieces of customer feedback returned by the mock API by changing the `NUM_CUSTOMERS` variable. 

#### Ingest customer feedback data into OpenSearch

> The first six tasks in the `analyze_customer_feedback` DAG perform the necessary steps to ingest data from the mock API into OpenSearch. 

Reviewer: ```suggestion
The first five tasks in the `analyze_customer_feedback` DAG perform the necessary steps to ingest data from the mock API into OpenSearch. 
```
I might have counted this wrong, pls double check my work!

    index_name=OPENSEARCH_INDEX_NAME,
    index_body={
        "settings": {
            "index": {
                "number_of_shards": 1,
                "knn": True,
                "knn.algo_param.ef_search": 100,
            }
        },
        "mappings": {
            "properties": {
                "customer_feedback": {"type": "text"},
                "customer_rating": {"type": "integer"},
                "customer_id": {"type": "keyword"},
                "timestamp": {"type": "date"},
                "customer_location": {"type": "keyword"},
                "product_type": {"type": "keyword"},
                "ab_test_group": {"type": "keyword"},
                "embeddings": {
                    "type": "knn_vector",
                    "dimension": MODEL_VECTOR_LENGTH,
                    "method": {
                        "name": "hnsw",
                        "space_type": "cosinesimil",
                        "engine": "nmslib",
                    },
                },
                "sentiment_prediction": {"type": "keyword"},
                "sentiment_confidence": {"type": "float"},
            }
        },
    },
)
```

> :::note

Reviewer: ```suggestion
:::info
```

            ),
            namespace=uuid.NAMESPACE_DNS,
        )
        kwargs = {"doc_id": str(unique_line_id), "document": customer}

        list_of_feedback.append(kwargs)

    return list_of_feedback
```

The resulting list of dictionaries takes the form of:

```python
[
    {
        "doc_id": "9c3880f4-f883-50d2-8a41-056c0efe4b2b",
        "document": {
            "ab_test_group": "A",
            "customer_feedback": "This product has transformed the way I work, absolutely fantastic UX!",
            "customer_id": 1714,
            "customer_location": "Switzerland",
            "customer_rating": 3,
            "product_type": "cloud service A",
            "timestamp": "2023-11-24T11:25:37Z",
        },
    },
    {
        "doc_id": "804dfcfd-62f5-5f8c-b64e-b167dcb4b1dc",
        "document": {
            "ab_test_group": "A",
            "customer_feedback": "The product did not meet my expectations, ugly UI.",
            "customer_id": 1843,
            "customer_location": "Switzerland",
            "customer_rating": 1,
            "product_type": "cloud service A",
            "timestamp": "2023-11-24T11:25:37Z",
        },
    },
]
```

> In the final stage of the ingestion part of the DAG, the `add_lines_as_documents` task uses the [OpenSearchAddDocumentOperator](https://registry.astronomer.io/providers/apache-airflow-providers-opensearch/versions/latest/modules/OpenSearchAddDocumentOperator) to add the customer feedback data to the OpenSearch index. This task is [dynamically mapped](dynamic-tasks.md) over the list of dictionaries returned by the `customer_feedback_to_dict_list` task to create one mapped task instance per set of `doc_id` and `document` parameter inputs. To map over a list of dictionaries, the `.expand_kwargs` method is used. 

Reviewer: ```suggestion
Lastly, the `add_lines_as_documents` task uses the [OpenSearchAddDocumentOperator](https://registry.astronomer.io/providers/apache-airflow-providers-opensearch/versions/latest/modules/OpenSearchAddDocumentOperator) to add the customer feedback data to the OpenSearch index. This task is [dynamically mapped](dynamic-tasks.md) over the list of dictionaries returned by the `customer_feedback_to_dict_list` task to create one mapped task instance per set of `doc_id` and `document` parameter inputs. To map over a list of dictionaries, the `.expand_kwargs` method is used. 
```

```python
add_lines_as_documents = OpenSearchAddDocumentOperator.partial(
    task_id="add_lines_as_documents",
    opensearch_conn_id=OPENSEARCH_CONN_ID,
    trigger_rule="none_failed",
    index_name=OPENSEARCH_INDEX_NAME,
).expand_kwargs(list_of_document_kwargs)
```

#### Query customer feedback data from OpenSearch

In the second part of the DAG, OpenSearch is queried to get the subset of customer feedback data we are interested in. For this example, we chose feedback from Swiss customers in the `A` test group using `cloud service A` who mentioned the user experience in their feedback.

![Graph view of the query section of the analyze_customer_feedback DAG.](/img/examples/use-case-airflow-cohere-opensearch_query_section.png)

> The `search_for_relevant_feedback` task uses the [OpenSearchQueryOperator](https://registry.astronomer.io/providers/apache-airflow-providers-opensearch/versions/latest/modules/OpenSearchQueryOperator) to query OpenSearch for the relevant customer feedback data. The query is defined as a dictionary and passed to the `query` parameter of the operator and will fuzzy match the terms provided in the `FEEDBACK_SEARCH_TERMS` variable while filtering for the `CUSTOMER_LOCATION`, `AB_TEST_GROUP`, and `PRODUCT_TYPE` variables.

Reviewer: ```suggestion
The `search_for_relevant_feedback` task uses the [OpenSearchQueryOperator](https://registry.astronomer.io/providers/apache-airflow-providers-opensearch/versions/latest/modules/OpenSearchQueryOperator) to query OpenSearch for the relevant customer feedback data. The query is defined as a dictionary and passed to the `query` parameter of the operator. OpenSearch fuzzy matches the terms provided in the `FEEDBACK_SEARCH_TERMS` variable while filtering for the `CUSTOMER_LOCATION`, `AB_TEST_GROUP`, and `PRODUCT_TYPE` variables.
```

```python
search_for_relevant_feedback = OpenSearchQueryOperator(
    task_id="search_for_relevant_feedback",
    opensearch_conn_id=OPENSEARCH_CONN_ID,
    index_name=OPENSEARCH_INDEX_NAME,
    query={
        "size": MAX_NUMBER_OF_RESULTS,
        "query": {
            "bool": {
                "must": [
                    {
                        "match": {
                            "customer_feedback": {
                                "query": FEEDBACK_SEARCH_TERMS,
                                "analyzer": "english",
                                "fuzziness": "AUTO",
                            }
                        }
                    }
                ],
                "filter": [
                    {"term": {"customer_location": CUSTOMER_LOCATION}},
                    {"term": {"ab_test_group": AB_TEST_GROUP}},
                    {"term": {"product_type": PRODUCT_TYPE}},
                ],
            },
        },
    },
)
```

> :::note

Reviewer: ```suggestion
:::info
```

> For more information on OpenSearch queries written in query domain-specific language (DSL), see the [OpenSearch documentation on Query DSL](https://opensearch.org/docs/latest/query-dsl/index/).

Reviewer: ```suggestion
For more information about OpenSearch queries written in query domain-specific language (DSL), see the [OpenSearch documentation on Query DSL](https://opensearch.org/docs/latest/query-dsl/index/).
```

        "product_type": "cloud service A",
        "timestamp": "2023-11-24T11:25:37Z",
        "id": "9c3880f4-f883-50d2-8a41-056c0efe4b2b",
    },
]
```

A second task, the `get_feedback_texts` is dynamically mapped over the list of dictionaries returned by the `reformat_relevant_reviews` task to extract the `customer_feedback` field from each dictionary. 

```python
@task
def get_feedback_texts(review_of_interest: dict) -> str:
    "Get the feedback text from the relevant reviews."
    feedback_text = review_of_interest["customer_feedback"]
    return feedback_text

feedback_texts = get_feedback_texts.expand(review_of_interest=relevant_reviews)
```

#### Perform sentiment analysis on relevant customer feedback and get embeddings using the Cohere API

The third section of the DAG consists of four tasks that perform sentiment analysis, get vector embeddings and load the results back into OpenSearch.

![Graph view of the section in the analyze_customer_feedback DAG that performs sentiment analysis, vector embeddings with the Cohere API and loads the results back into OpenSearch.](/img/examples/use-case-airflow-cohere-opensearch_sentiment_embedding_section.png)

> The first task in this section, `get_sentiment` uses the [CohereHook](https://registry.astronomer.io/providers/cohere/versions/latest/modules/CohereHook) to get the sentiment of the customer feedback using the Cohere API [Text classification endpoint](https://docs.cohere.com/reference/classify). The task is dynamically mapped over the list of feedback texts returned by the `get_feedback_texts` task to create one mapped task instance per customer feedback to be analyzed in parallel. Sentiment examples are stored in the [`classification_examples` file](https://github.com/astronomer/use-case-llm-customer-feedback/blob/main/include/classification_examples.py) in the `include` folder. 

Reviewer: ```suggestion
The first task in this section, `get_sentiment`, uses the [CohereHook](https://registry.astronomer.io/providers/cohere/versions/latest/modules/CohereHook) to get the sentiment of the customer feedback using the Cohere API [Text classification endpoint](https://docs.cohere.com/reference/classify). The task is dynamically mapped over the list of feedback texts returned by the `get_feedback_texts` task to create one mapped task instance per customer feedback to be analyzed in parallel. Sentiment examples are stored in the [`classification_examples` file](https://github.com/astronomer/use-case-llm-customer-feedback/blob/main/include/classification_examples.py) in the `include` folder. 
```

First, the `get_embeddings_testimonial_search_term` task converts the target testimonial to vector embeddings using the [CohereEmbeddingOperator](https://registry.astronomer.io/providers/cohere/versions/latest/modules/CohereEmbeddingOperator). 

```python
get_embeddings_testimonial_search_term = CohereEmbeddingOperator(
    task_id="get_embeddings_testimonial_search_term",
    conn_id=COHERE_CONN_ID,
    input_text=TESTIMONIAL_SEARCH_TERM,
)
```

Next, the `prep_search_term_embeddings_for_query` task converts the embeddings returned by the `get_embeddings_testimonial_search_term` task to a list of floats to be used in the OpenSearch query.

```python
@task
def prep_search_term_embeddings_for_query(embeddings: list) -> list:
    "Prepare the embeddings for the OpenSearch query."
    return [float(x) for x in embeddings[0]]

search_term_embeddings = prep_search_term_embeddings_for_query(
    embeddings=get_embeddings_testimonial_search_term.output
)
```

> The `search_for_testimonial_candidates` task uses the [OpenSearchQueryOperator](https://registry.astronomer.io/providers/apache-airflow-providers-opensearch/versions/latest/modules/OpenSearchQueryOperator) to query OpenSearch for the most similar customer feedback to the target testimonial using a k-NN algorithm on the embeddings and filter for positive sentiment. Note that k-NN search requires the `knn` plugin to be installed in OpenSearch.

Reviewer: ```suggestion
The `search_for_testimonial_candidates` task uses the [OpenSearchQueryOperator](https://registry.astronomer.io/providers/apache-airflow-providers-opensearch/versions/latest/modules/OpenSearchQueryOperator) to query OpenSearch for the most similar customer feedback to the target testimonial using a [k-NN](https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/)) algorithm on the embeddings and filter for positive sentiment. Note that k-NN search requires the `knn` plugin to be installed in OpenSearch.
```
See note below

```python
search_for_testimonial_candidates = OpenSearchQueryOperator(
    task_id="search_for_testimonial_candidates",
    opensearch_conn_id=OPENSEARCH_CONN_ID,
    index_name=OPENSEARCH_INDEX_NAME,
    query={
        "size": 10,
        "query": {
            "bool": {
                "must": [
                    {
                        "knn": {
                            "embeddings": {
                                "vector": search_term_embeddings,
                                "k": 10,
                            }
                        }
                    }
                ],
                "filter": [
                    {"term": {"sentiment_prediction": "positive"}},
                ],
            }
        },
    },
)
```

> :::note

Reviewer: ```suggestion
:::info
```

Reviewer: Looking at the note content, I wonder if it would be better to simply hyperlink the initial instance of `k-NN` instead of having a complete info note. It's a lot of real estate for a docs link, and I typically associate notes with something we really think the user should know or check out. Having this note without a specific call to action or consideration feels like a possible distraction. If you agree, we can also change the other notes in the guide to fit that format.

Author: Yes, good callout! I'll change the notes :)

For more information on using k-NN search in OpenSearch, see the [OpenSearch documentation on k-NN search](https://opensearch.org/docs/latest/search-plugins/knn/filter-search-knn/).

:::

Lastly, the `print_testimonial_candidates` task prints positive customer feedback that is closest to the target testimonial feedback and mentions the user experience of the `cloud A service` to the logs.

```python
@task
def print_testimonial_candidates(search_results: dict) -> None:
    "Print the testimonial candidates from the OpenSearch query results."
    for result in search_results["hits"]["hits"]:
        print("Customer ID: ", result["_source"]["customer_id"])
        print("Customer feedback: ", result["_source"]["customer_feedback"])
        print("Customer location: ", result["_source"]["customer_location"])
        print("Customer rating: ", result["_source"]["customer_rating"])
        print("Customer sentiment: ", result["_source"]["sentiment_prediction"])
        print(
            "Customer sentiment confidence: ",
            result["_source"]["sentiment_confidence"],
        )
```

> You can review the output of the task in the task logs, they should look similar to the following:

Reviewer: ```suggestion
You can review the output of the task in the task logs. They should look similar to the following:
```
